grid_map_visualization:
  ros__parameters:
    grid_map_topic: /costmap
    grid_map_visualizations: [slope, traversability, traversability_inflated]

    slope:
      type: occupancy_grid
      params:
        layer: slope
        data_min: 0.0
        data_max: 1.57

    traversability:
      type: occupancy_grid
      params:
        layer: traversability
        data_min: 0.0
        data_max: 1.0

    traversability_inflated:
      type: occupancy_grid
      params:
        layer: traversability_inflated
        data_min: 0.0
        data_max: 1.0

traversability_mapping:
  ros__parameters:
    pointcloud_topic: /cloud_unstructured_fullframe
    output_map_topic: /costmap
    filter_radius: 5.0

    filters:
      filter1:
        name: buffer_normalizer
        type: gridMapFilters/BufferNormalizerFilter

      filter2: # Reduce noise with a radial blurring filter.
        name: mean_in_radius
        type: gridMapFilters/MeanInRadiusFilter
        params:
          input_layer: elevation
          output_layer: elevation_smooth
          radius: 0.1

      filter3: # Compute surface normals.
        name: surface_normals
        type: gridMapFilters/NormalVectorsFilter
        params:
          input_layer: elevation
          output_layers_prefix: normal_vectors_
          radius: 0.5
          normal_vector_positive_axis: z

      filter4: # Compute slope from surface normal.
        name: slope
        type: gridMapFilters/MathExpressionFilter
        params:
          output_layer: slope
          expression: acos(normal_vectors_z)

      filter5: # Zero out slope values below a threshold.
        name: slope_lower_threshold
        type: gridMapFilters/ThresholdFilter
        params:
          layer: slope
          lower_threshold: 1.0
          set_to: 0.0

      filter6: # Compute roughness as absolute difference from map to smoothened map.
        name: roughness
        type: gridMapFilters/MathExpressionFilter
        params:
          output_layer: roughness
          expression: abs(elevation - elevation_smooth)

      filter7: # Compute traversability as normalized weighted sum of slope and roughness.
        name: traversability
        type: gridMapFilters/MathExpressionFilter
        params:
          output_layer: traversability
          expression: slope / 1.57

      filter8: # Apply a gaussian kernel with a given radius and coefficient.
        name: gaussian_kernel
        type: urcPerception/GaussianFilter
        params:
          input_layer: traversability
          output_layer: traversability_inflated
          radius: 0.5
          gaussian_coeff: 4.0
